language/src/main/java/com/vztekoverflow/cilostazol/runtime/symbols/TypeSymbol.java

```diff
@@ -1,26 +1,122 @@
 package com.vztekoverflow.cilostazol.runtime.symbols;
 
+import com.oracle.truffle.api.CompilerAsserts;
+import com.oracle.truffle.api.CompilerDirectives;
+import com.oracle.truffle.api.nodes.ExplodeLoop;
 import com.vztekoverflow.cil.parser.cli.signature.SignatureReader;
 import com.vztekoverflow.cil.parser.cli.signature.TypeSig;
 import com.vztekoverflow.cil.parser.cli.table.CLITablePtr;
 import com.vztekoverflow.cil.parser.cli.table.generated.CLITableConstants;
 import com.vztekoverflow.cil.parser.cli.table.generated.CLITypeSpecTableRow;
 import com.vztekoverflow.cilostazol.CILOSTAZOLBundle;
+import com.vztekoverflow.cilostazol.meta.SystemTypes;
 import com.vztekoverflow.cilostazol.runtime.other.ContextProviderImpl;
 import com.vztekoverflow.cilostazol.runtime.typesystem.TypeSystemException;
 
 public abstract class TypeSymbol extends Symbol {
+  private final SystemTypes kind;
   protected final ModuleSymbol definingModule;
 
   public TypeSymbol(ModuleSymbol definingModule) {
     super(ContextProviderImpl.getInstance());
     this.definingModule = definingModule;
+    // TODO
+    this.kind = SystemTypes.Object;
   }
 
   public ModuleSymbol getDefiningModule() {
     return definingModule;
   }
 
+  public boolean isInterface() {
+    return false;
+  }
+
+  public boolean isArray() {
+    return false;
+  }
+
+  public NamedTypeSymbol[] getInterfaces() {
+    return new NamedTypeSymbol[0];
+  }
+
+  public NamedTypeSymbol[] getSuperTypes() {
+    return new NamedTypeSymbol[0];
+  }
+
+  protected int getHierarchyDepth() {
+    // TODO
+    return 0;
+  }
+
+  public SystemTypes getKind() {
+    return kind;
+  }
+
+  public boolean isAssignableFrom(TypeSymbol other) {
+    if (this == other) return true;
+
+    if (this.isArray()) {
+      if (other.isArray()) {
+        TODO:
+        return false; // ((ArrayKlass) this).arrayTypeChecks((ArrayKlass) other);
+      }
+    }
+
+    if (this.isInterface()) {
+      return checkInterfaceSubclassing(other);
+    }
+    return checkOrdinaryClassSubclassing(other);
+  }
+
+  /**
+   * Performs type checking for non-interface, non-array classes.
+   *
+   * @param other the class whose type is to be checked against {@code this}
+   * @return true if {@code other} is a subclass of {@code this}
+   */
+  public boolean checkOrdinaryClassSubclassing(TypeSymbol other) {
+    int depth = getHierarchyDepth();
+    return other.getHierarchyDepth() >= depth && other.getSuperTypes()[depth] == this;
+  }
+
+  /**
+   * Performs type checking for interface classes.
+   *
+   * @param other the class whose type is to be checked against {@code this}
+   * @return true if {@code this} is a super interface of {@code other}
+   */
+  public boolean checkInterfaceSubclassing(TypeSymbol other) {
+    // TODO: Are these interfaces transitive?
```

Comment:
No, you only get direct inherited interfaces, but not interfaces of direct base class for example.